RSDK-7753: Remove non-blocking GoFor functionality on 9/1

martha-johnston opened the pull request:
Removed revolutions == 0 functionality from all the motors. 

Now I'm wondering if there should be a check that errors if you pass in 0 revolutions? As I was going through I checked there would be no divided by zero issues, and it seems fine (but let me know if I missed any). I don't feel strongly about adding an error for every GoFor but I see the benefit of it so let me know what you think.

randhid commented on components/motor/motor.go:
```diff
@@ -104,8 +104,6 @@ type Motor interface {
 	// revolutions at a given speed in revolutions per minute. Both the RPM and the revolutions
 	// can be assigned negative values to move in a backwards direction. Note: if both are
 	// negative the motor will spin in the forward direction.
-	// If revolutions != 0, this will block until the number of revolutions has been completed or another operation comes in.
-	// Deprecated: If revolutions is 0, this will run the motor at rpm indefinitely.
```
Keep the note about blocking, that's important information about this API and part of our API contract.

randhid commented on components/motor/gpio/mathutils.go:
```diff
@@ -34,11 +32,6 @@ func goForMath(maxRPM, rpm, revolutions float64) (float64, time.Duration) {
-		return powerPct, 0
-	}
-
 	dir := rpm * revolutions / math.Abs(revolutions*rpm)
```
There's a divide by zero case that we need to test for now returning from goForMath.

martha-johnston replied:
oops missed a lot

randhid replied:
Yeah.... run `make test-go` locally or `go test` in the folders you've edited so you realize before you get to the PR phase, it always helps and I see a lot of the math logic is failing, so that's great!

randhid commented on components/motor/gpio/mathutils.go:
```diff
@@ -48,17 +41,10 @@ func goForMath(maxRPM, rpm, revolutions float64) (float64, time.Duration) {
-		direction = sign(rpm)
-	}
 
 	goalPos := (math.Abs(revolutions) * ticksPerRotation * direction) + currentPos
```
zero rev forces this to currentPos, which is great.

randhid commented on components/motor/gpio/encoded.go:
```diff
@@ -239,8 +239,6 @@ func (m *EncodedMotor) SetPower(ctx context.Context, powerPct float64, extra map
 // revolutions at a given speed in revolutions per minute. Both the RPM and the revolutions
 // can be assigned negative values to move in a backwards direction. Note: if both are
 // negative the motor will spin in the forward direction.
-// If revolutions != 0, this will block until the number of revolutions has been completed or another operation comes in.
```
keep this part of the comment.

randhid commented on components/motor/fake/motor.go:
```diff
@@ -264,11 +262,6 @@ func goForMath(maxRPM, rpm, revolutions float64) (float64, time.Duration, float6
-		return powerPct, 0, 1
-	}
-
 	dir := rpm * revolutions / math.Abs(revolutions*rpm)
```
divide by zero case, can you test what happens with a fake motor?

randhid commented on components/motor/dimensionengineering/sabertooth.go:
```diff
@@ -556,11 +549,6 @@ func goForMath(maxRPM, rpm, revolutions float64) (float64, time.Duration) {
-		return powerPct, 0
-	}
-
 	dir := rpm * revolutions / math.Abs(revolutions*rpm)
```
divide by zero case.

martha-johnston commented on components/motor/dimensionengineering/sabertooth_test.go:
```diff
@@ -1,15 +1,16 @@
-package dimensionengineering_test
+package dimensionengineering
```
I changed this package to easier test goForMath because a lot of other motor packages don't use this `_test` convention, but I am open to changing it back if there's a good reason to do so. It's not fully clear to me what this is useful for.

randhid replied:
This is actually a convention for making sure your test file is in a separate package which forces you to write good tests of the functionality of your exported APIs from that package. If you don't need any of the private functions in the dimensionengineering package to be tested now, please change it back. 

https://stackoverflow.com/questions/19998250/proper-package-naming-for-testing-with-the-go-language/31443271#31443271

martha-johnston replied:
Sure I can revert this and remove the goForMath tests if you'd prefer

randhid replied:
> If you don't need any of the private functions in the dimensionengineering package to be tested now, please change it back.

You are testing private function now, so don't change it back.

martha-johnston replied:
sorry I missed this before changing it, but I added GoFor tests so I think it should be fine now

randhid replied:
Even better!

JohnN193 commented on components/motor/fake/motor.go:
```diff
@@ -251,25 +251,25 @@ func (m *Motor) Direction() int {
 
-// If revolutions is 0, the returned wait duration will be 0 representing that
-// the motor should run indefinitely.
 func goForMath(maxRPM, rpm, revolutions float64) (float64, time.Duration, float64) {
 	// need to do this so time is reasonable
 	if rpm > maxRPM {
 		rpm = maxRPM
 	} else if rpm < -1*maxRPM {
 		rpm = -1 * maxRPM
 	}
```
so since having `rpm = 0` or `revolutions = 0` are now invalid requests, maybe we can do something
```
func goForMath(maxRPM, rpm, revolutions float64) (float64, time.Duration, float64) {
        dir := 1.0
        
	// check that a valid request has been made
	if rpm == 0 || revolutions == 0{
	return 0, 0, dir
	}

        // need to do this so time is reasonable
	if rpm > maxRPM {
		rpm = maxRPM
	} else if rpm < -1*maxRPM {
		rpm = -1 * maxRPM
	}
	
	// flip sign if needed
	if rpm*revolutions < 0 {
	     dir = -1.0
	}
	
	powerPct := math.Abs(rpm) / maxRPM * dir
	waitDur := time.Duration(math.Abs(revolutions/rpm)*60*1000) * time.Millisecond
	return powerPct, waitDur, dir
}
```

JohnN193 replied:
we can also just return during the dir = 0 case instead, to have the same behavior

martha-johnston replied:
our pattern before this was to do those checks in GoFor, so I added a revolutions check to the exiting sped check but it's just not in goForMath

JohnN193 commented on components/motor/gpio/basic.go:
```diff
@@ -284,11 +284,6 @@ func (m *Motor) GoFor(ctx context.Context, rpm, revolutions float64, extra map[s
 		return errors.Wrap(err, "error in GoFor")
 	}
 
-	if revolutions == 0 {
```
just to confirm, we are choosing not to error when revolutions is 0, but rahter we will return a value from goForMath that results in the motors not moving

martha-johnston replied:
now we are erroring

JohnN193 commented on components/motor/gpio/mathutils.go:
```diff
@@ -34,12 +32,7 @@ func goForMath(maxRPM, rpm, revolutions float64) (float64, time.Duration) {
-	dir := rpm * revolutions / math.Abs(revolutions*rpm)
+	dir := sign(rpm * revolutions)
 	powerPct := math.Abs(rpm) / maxRPM * dir
 	waitDur := time.Duration(math.Abs(revolutions/rpm)*60*1000) * time.Millisecond
```
looks like we do not disallow 0 rpm here as well.

JohnN193 commented on components/motor/gpio/mathutils.go:
```diff
@@ -34,12 +32,7 @@ func goForMath(maxRPM, rpm, revolutions float64) (float64, time.Duration) {
-	}
-
-	dir := rpm * revolutions / math.Abs(revolutions*rpm)
+	dir := sign(rpm * revolutions)
```
lets return early `if dir == 0`, so we cover both 0 revolutions and 0 rpm cases

randhid replied:
I think we just need to check this logic everywhere and return early with an error, it's the simplest form of the rpm or revolution being zero check.

martha-johnston replied:
agreed, now the speed and revs are checked everywhere before this

JohnN193 commented on components/motor/gpio/mathutils.go:
```diff
@@ -48,17 +41,10 @@ func goForMath(maxRPM, rpm, revolutions float64) (float64, time.Duration) {
 // goForMath calculates goalPos, goalRPM, and direction based on the given GoFor rpm and revolutions, and the current position.
 func encodedGoForMath(rpm, revolutions, currentPos, ticksPerRotation float64) (float64, float64, float64) {
 	direction := sign(rpm * revolutions)
```
we could return early when direction == 0 here as well, but the logic below already handles the `rpm == 0` and `revolutions == 0` case, so maybe fine as is

JohnN193 commented on components/motor/roboclaw/roboclaw.go:
```diff
@@ -239,14 +239,13 @@ func (m *roboclawMotor) SetPower(ctx context.Context, powerPct float64, extra ma
-		return powerPct, 0
+	dir := 1.0
+	if rpm*revolutions == 0.0 {
+		dir = 0.0
```
similar recommendation, we can return early here, especially because the logic below breaks when `rpm == 0`

JohnN193 reviewed (commented):
seems like a bunch of GoFor cases do not handle the case when rpm is set to 0. I think we should handle that in this pr since we are already touching the GoForMath function

randhid commented on components/motor/fake/motor.go:
```diff
@@ -251,25 +251,25 @@ func (m *Motor) Direction() int {
+	dir := 1.0
+	if rpm*revolutions == 0.0 {
+		dir = 0.0
+	} else if rpm*revolutions < 0.0 {
+		dir = -1.0
 	}
 
-	if revolutions == 0 {
-		powerPct := rpm / maxRPM
-		return powerPct, 0, 1
+	if rpm == 0 {
+		return 0, 0, dir
```
let's make the checking functions helpers so we don't have to update them everywhere, we already have Errors and error helpers in motor.go, the check for 0 rpm and 0 revolutions is now the same., so we shouldn't copy/past identical code.

randhid replied:
This comment was not addressed. 

EDIT: I was not re-requested for review, you only committed, sorry! Was in a github PR review zone this morning.

randhid reviewed (changes requested):
Let's go a bit further and standardize the rpm and revolution checks for zero and return early with an error in GoFor. I think that's cleaner code-wise.

martha-johnston commented on components/motor/dimensionengineering/sabertooth.go:
```diff
@@ -556,12 +549,13 @@ func goForMath(maxRPM, rpm, revolutions float64) (float64, time.Duration) {
-		powerPct := rpm / maxRPM
-		return powerPct, 0
+	dir := 1.0
+	if rpm*revolutions == 0.0 {
+		dir = 0.0
```
seems fine to me? but now I did add a check to gofor, which is before this point, so we shouldn't get here anyways, but I thought I should leave the 0 case in case goForMath is called directly somewhere else at a future point in time?

martha-johnston commented on components/motor/dimensionengineering/sabertooth.go:
```diff
@@ -556,12 +549,13 @@ func goForMath(maxRPM, rpm, revolutions float64) (float64, time.Duration) {
 
-	dir := rpm * revolutions / math.Abs(revolutions*rpm)
 	powerPct := math.Abs(rpm) / maxRPM * dir
 	waitDur := time.Duration(math.Abs(revolutions/rpm)*60*1000) * time.Millisecond
```
in goFor rpm of 0 would error, and now revolutions of 0 does as well

martha-johnston commented:
I added a new check for if the revolutions is 0, and I added a new helper for getting the direction to reduce repeated code. I think this covers a lot of the similar comments but I'll go through it again and let me know if there's more

randhid commented on components/motor/motor.go:
```diff
@@ -200,3 +199,22 @@ func CheckSpeed(rpm, max float64) (string, error) {
+	return nil
+}
+
+// GetDirection returns the direction based on the rpm and revolutions
+func GetDirection(rpm, revolutions float64) float64 {
+	dir := 1.0
+	if rpm*revolutions == 0.0 {
+		dir = 0.0
+	} else if rpm*revolutions < 0.0 {
+		dir = -1.0
+	}
+	return dir
+}
```
Please rename this because `GetDirection` implies I'm getting the motor's direction as it is is moving, not getting the commanded Direction from the rpm and revolution requests.

JohnN193 reviewed (approved):
kk still not the biggest fan of having math/codepaths within `goForMath`s that could potentially give invalid outputs, but it shouldn't be an issue that users hit. 

maybe at a future date we could try unifying goForMath functions into a generic set of helpers? like adding a GoForMathDuration and GoForMathEncoded helpers, that way we won't need to change every function in drivers we support

randhid commented on components/motor/fake/motor.go:
```diff
@@ -351,8 +343,9 @@ func (m *Motor) GoTo(ctx context.Context, rpm, pos float64, extra map[string]int
 	if err != nil {
 		return err
 	}
-	if curPos == pos {
-		return nil
```
Is this changing behaviour? If it was zero previously, it would just return from GoTo, and now it errors. The response change changes the behaviour when this is used in the client code or another driver. 

THis is a fake motor so not super important, but always keep an eye out for this sort of change, it is breaking functionality, if not our own API.

randhid replied:
Ah no, now I see the err != nil check.

randhid reviewed (approved):
LGTM now!
